### PR TITLE
Allocate all groups in one go

### PR DESCRIPTION
**Pull request by bratseth**
With many groups and dynamic allocation allocating group by group is too slow.

This leaves a number of simplification opportunities untaken, but I wanted to keep this diff as small as possible.

### CODE REVIEW

**Comment by hmusum** [comment]
I am unable to review all of this today, so just a comment on of the changes: 

Cluster membership format is changed with this PR. Won't there be problems when upgrading, since the old code won't be able to read the new format? In that case, code for supporting the new cluster membership format needs to go in first, then the rest of the changes can be done.

**Comment by bratseth** [comment]
I'm pretty sure it's ok since the specs without group were never returned out of the node repo before this PR - and could not be parsed at all since the constructor started with

```
        if (components.length < 4)
            throw new RuntimeException("Could not parse '" + stringValue + "' to a cluster membership. " +
                                       "Expected 'clusterType/clusterId/groupId/index[/retired][/exclusive][/stateful][/combinedId]'");

```

File: config-provisioning/src/main/java/com/yahoo/config/provision/ClusterMembership.java
Comment (hmusum) [inline config-provisioning/src/main/java/com/yahoo/config/provision/ClusterMembership.java:25]:
```suggestion
        if (components.length < 4)
```

**Comment by hmusum** [comment]
Yes, right, that looks good

File: config-provisioning/src/main/java/com/yahoo/config/provision/ClusterMembership.java
Comment (bratseth) [reply config-provisioning/src/main/java/com/yahoo/config/provision/ClusterMembership.java:25]: I wanted to support everything this can output now ...

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/NodeAllocation.java
Comment (freva) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/NodeAllocation.java:304]: Not sure I understand this... Seems like we want to return true if the group the given `node` is in already has more nodes than what is requested, then this node can instead by reallocated to a different group? If so, then shouldn't the sign be reversed?

Comment (bratseth) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/NodeAllocation.java:304]: The premise is: We cannot _change_ the group of a node we will keep (unless that group is to be removed), because that will lead to temporary coverage loss.

This method returns whether - given that - this node can be used in a group which has to few nodes. If the group the node belongs to has fewer nodes than it should, it can.

File: node-repository/src/test/java/com/yahoo/vespa/hosted/provision/NodeListMicroBenchmarkTest.java
Comment (mpolden) [inline node-repository/src/test/java/com/yahoo/vespa/hosted/provision/NodeListMicroBenchmarkTest.java:61]: This is the whole point of this (ignored) test, so it shouldn't be removed.

**Review by hmusum — DISMISSED** [review]
Looks good to me, but a lot of changes and high risk, so maybe someone else should have a look as well

File: config-provisioning/src/main/java/com/yahoo/config/provision/ClusterMembership.java
Comment (jonmv) [reply config-provisioning/src/main/java/com/yahoo/config/provision/ClusterMembership.java:25]: So ideally, this should roll out first, since upgrades aren't atomic.

Comment (jonmv) [reply config-provisioning/src/main/java/com/yahoo/config/provision/ClusterMembership.java:25]: Ah, well, no, the converse to this should roll out first, i.e, when an upgraded server writes a spec with `//`, the other servers should still be able to handle that meaningfully. They won't now, will they?

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/GroupIndices.java
Comment (jonmv) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/GroupIndices.java:145]: Isn't it necessary to check the free parent capacity here, as well?

Comment (jonmv) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/GroupIndices.java:145]: I see that wasn't done before, so could do separately, if needed.

Comment (bratseth) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/GroupIndices.java:145]: We don't create candidates on hosts where there isn't capacity.

**Comment by bratseth** [comment]
@jonmv, @freva, @mpolden let me know if any of you want to look more at this before I merge.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/GroupIndices.java
Comment (jonmv) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/GroupIndices.java:145]: Ah, that's good.

**Comment by jonmv** [comment]
I don't want to look more at it, unless you need to let old configservers handle the format written with `//` by upgraded servers with the new cluster membership format ... ?

File: config-provisioning/src/main/java/com/yahoo/config/provision/ClusterMembership.java
Comment (jonmv) [reply config-provisioning/src/main/java/com/yahoo/config/provision/ClusterMembership.java:25]: Here.

**Comment by bratseth** [comment]
... turns out it was easy to move storing of the reserved nodes to after assigning groups so I just did that.

**Review by mpolden — APPROVED** [review]
LGTM. Hopefully there are integration tests that use multiple groups?

**Comment by bratseth** [comment]
There are.

**Pull request merged**
